### DISCUSSION TITLE
Unsealing Builders and Configuration

### DISCUSSION THREAD

**Discussion by pinkfloydx33** [testcontainers/testcontainers-dotnet#829] [General]
Any chance with the new module-specific builders/configuration/etc. in 3.0.0 that these types will get unsealed? It would be nice to override/extend from them to create our own customized versions that we can use throughout out projects. For example we always override the postgres image and user name; it'd be nice if we could have our own module without have to re-create it entirely from scratch.

The [documentation](https://dotnet.testcontainers.org/modules/#extend-a-module) makes it sound like you should be able to extend (inherit) from the builder and configuration classes, however everything is sealed.

**Comment by HofmeisterAn** [comment]
Yes, that is something we briefly discussed and have been giving it a lot of thought (and still do). While we would like to provide a pre-configured instance that is ready to use upon creation, including support for inheritance, developers who extend modules may encounter a null reference exception due to a virtual member call.

To prevent developers from encountering this issue, we have decided to keep the builders sealed for now until we can come up with a better design. We are open to discussing proposals and all ideas that could improve the design. I have a few ideas in mind, but I need to test them first.

> For example we always override the postgres image and user name; it'd be nice if we could have our own module without have to re-create it entirely from scratch.

Additionally, are these the only use cases for which you would like to unseal the classes? I ask because I would like to understand if there are other cases we may be overlooking.

If you need multiple instances of the same or similar configurations, you can call `Build()` multiple times. The builder holds an immutable instance of the configuration, and calling `Build()` creates a new container instance each time. Therefore, you can share builders:

```csharp
var sharedBuilder = new ContainerBuilder()
    .WithImage(CommonImages.Alpine)
    .WithEntrypoint(CommonCommands.SleepInfinity);

var containerA = sharedBuilder
    .WithEnvironment("DOTNET_NOLOGO", "true")
    .Build();

var containerB = sharedBuilder
    .WithEnvironment("DOTNET_NOLOGO", "false")
    .Build();

var containerC = sharedBuilder.Build();
var containerD = sharedBuilder.Build();
var containerE = sharedBuilder.Build();
```

**Reply by dimabarbul** [reply]
Hello

My use case is that I want to do some setup before/after container starts. For example, when RabbitMQ container starts, I want to setup permissions and exchanges.

Now I see following options:
1) I can do it using rabbitmq.config+definitions.json
but this way is pretty complex as I need to generate those files and then provide guidance on how to regenerate them if they need to be updated
2) create entrypoint.sh and put logic there
I'm not sure this option can be used in this particular scenario (as we need to start rabbitmq service, wait for port to be available and only after that do the setup), but maybe I need to dig deeper to make it work
3) inherit from RabbitMqConfiguration to add setup configuration, extend RabbitMqBuilder to add methods for filling the configuration, extend RabbitMqContainer to do setup based on the configuration in StartAsync
4) Copy RabbitMq* classes and add my logic
for the moment, the best option I have, but I'd like to base my work on top of existing package providing basic features, like setting up username/password in builder, providing connection string in container etc.

The third option seems to be much simpler and more reliable to me.

**Reply by HofmeisterAn** [reply]
Maybe there are a few other things you can try:

- Use a template configuration file to generate the config (or replace specific parts) at runtime. Then, upload the file before the container starts using the [`WithResourceMapping(string, string)`](https://dotnet.testcontainers.org/api/create_docker_container/#copying-directories-or-files-to-the-container) API (depending on the configuration file, this is likely the simplest approach and how I would start).
- The container builder provides an API that lets you run a callback right after the container starts. For example, in [Kafka](https://github.com/testcontainers/testcontainers-dotnet/blob/6526e5a24753e3e2a32beb844760d67d6c5bff2f/src/Testcontainers.Kafka/KafkaBuilder.cs#L117-L138) (see also L:115), we use this to modify the entrypoint.
- The container type triggers a few [events](https://github.com/testcontainers/testcontainers-dotnet/blob/6526e5a24753e3e2a32beb844760d67d6c5bff2f/src/Testcontainers/Containers/IContainer.cs#L19-L77) during its lifecycle. If you go this route, make sure to handle synchronization properly.

**Reply by dimabarbul** [reply]
Thanks for the quick response.

Let me add some more information. There is a number of microservices (I'll call them client code) where I'd like to add integration tests using Testcontainers. In order to make it easier to consume in the client code I'd like to pack custom Testcontainers in nuget and share between them. Some of the dependencies require setup, so making custom builder-configuration-container classes seems the best approach. So ideal solution for me would lead to approximately following in the client code:

```csharp
var rabbitMq = new RabbitMqBuilder() // my custom class
    .WithUsername("guest") // method from Testcontainers.RabbitMq
    .WithPassword("guest") // method from Testcontainers.RabbitMq
    .WithAllMightyUser("admin", "password") // custom method
    .WithReadOnlyUser("ro-user", "password", "vhost1", "vhost2") // custom method
    .Build();
await rabbitMq.StartAsync();
var defaultCS = rabbitMq.GetConnectionString(); // method from Testcontainers.RabbitMq
var roCS = rabbitMq.GetConnectionString("ro-user"); // custom method
```

The `RabbitMqBuilder` from the code above is my custom builder class. I need it to 1) encapsulate all the code needed for settings up permissions from client code, which 2) makes client code much easier to incorporate future improvements - they'll need to update package version. Apparently, custom builder class requires custom `RabbitMqConfiguration` and `RabbitMqContainer` classes.
And here's the problem - as I cannot inherit from `Testcontainers.RabbitMq.RabbitMqBuilder`, I cannot reuse any of the code (for example, setting default username/password, or wait strategy for rabbitmq container), I have to copy it and later on monitor changes in next versions and copy them manually.
This essentially means, I will use `Testcontainers.RabbitMq` as an example and not as dependency.

And in general, I want to add more functionality to classes from `Testcontainers.RabbitMq` package with some modifications, which in OOP means inheritance (aggregation is not possible here as many of required constructors and methods are not public). But instead of taking this IMHO straightforward way and get all the benefits of building my work on top of OSS (for example, when wait strategy changes, I don't need to check source code, I just need to update my package dependency), I need to look for some workarounds.

Having all that said, IMHO, the easiest and more "natural" way of achieving flexibility along with reliability would be to support inheritance.

**Reply by dimabarbul** [reply]
> Maybe there are a few other things you can try:
> 
>     * Use a template configuration file to generate the config (or replace specific parts) at runtime. Then, upload the file before the container starts using the [`WithResourceMapping(string, string)`](https://dotnet.testcontainers.org/api/create_docker_container/#copying-directories-or-files-to-the-container) API (depending on the configuration file, this is likely the simplest approach and how I would start).
> 
>     * The container builder provides an API that lets you run a callback right after the container starts. For example, in [Kafka](https://github.com/testcontainers/testcontainers-dotnet/blob/6526e5a24753e3e2a32beb844760d67d6c5bff2f/src/Testcontainers.Kafka/KafkaBuilder.cs#L117-L138) (see also L:115), we use this to modify the entrypoint.
> 
>     * The container type triggers a few [events](https://github.com/testcontainers/testcontainers-dotnet/blob/6526e5a24753e3e2a32beb844760d67d6c5bff2f/src/Testcontainers/Containers/IContainer.cs#L19-L77) during its lifecycle. If you go this route, make sure to handle synchronization properly.

As for the suggestions:

1) I don't really want to mess with RabbitMq config file, because it's quite a big burden - generate template, watch for it updates in new RabbitMq versions, provide guidance for updating it etc.

2) Startup callback does not work in this case as creating exchanges and permissions should happen after the container is ready, but startup callback is triggered right after the container is started, i.e. without waiting for the server to be able to handle the commands.

3) Event is triggered after the container is ready, but creating permissions and exchanges is a part of making container ready, so it should happen before anyone can start using it.
